File: LR1/DoublyList.cpp
#include "DoublyList.h"
#include <iostream>
#include <fstream>

using namespace std;

void LADD_HEAD(DoublyList& list, const string& value) {
    LNode* newNode = new LNode(value, list.head, nullptr);
    
    if (!list.head) {
        list.head = list.tail = newNode;
    } else {
        newNode->next = list.head;
        list.head->prev = newNode;
        list.head = newNode;
    }
    list.size++;
}

void LADD_TAIL(DoublyList& list, const string& value) {
    LNode* newNode = new LNode(value, nullptr, list.tail);
    
    if (!list.head) {
        list.head = list.tail = newNode;
    } else {
        newNode->prev = list.tail;
        list.tail->next = newNode;
        list.tail = newNode;
    }
    list.size++;
}

void LADD_BEFORE(DoublyList& list, const string& value, const string& target) {
    if (!list.head) {
        cout << "Список пуст!\n";
        return;
    }
    
    LNode* curr = list.head;
    while (curr && curr->data != target) {
        curr = curr->next;
    }
    
    if (!curr) {
        cout << "Элемент " << target << " не найден!\n";
        return;
    }
    
    LNode* newNode = new LNode(value, curr, curr->prev);
    
    if (curr == list.head) {
        newNode->next = list.head;
        list.head->prev = newNode;
        list.head = newNode;
    } else {
        newNode->prev = curr->prev;
        newNode->next = curr;
        curr->prev->next = newNode;
        curr->prev = newNode;
    }
    list.size++;
}

void LADD_AFTER(DoublyList& list, const string& value, const string& target) {
    if (!list.head) {
        cout << "Список пуст!\n";
        return;
    }
    
    LNode* curr = list.head;
    while (curr && curr->data != target) {
        curr = curr->next;
    }
    
    if (!curr) {
        cout << "Элемент " << target << " не найден!\n";
        return;
    }
    
    LNode* newNode = new LNode(value, curr->next, curr);
    
    if (curr == list.tail) {
        newNode->prev = list.tail;
        list.tail->next = newNode;
        list.tail = newNode;
    } else {
        newNode->prev = curr;
        newNode->next = curr->next;
        curr->next->prev = newNode;
        curr->next = newNode;
    }
    list.size++;
}

void LREMOVE_HEAD(DoublyList& list) {
    if (!list.head) {
        cout << "Список пуст!\n";
        return;
    }
    
    LNode* temp = list.head;
    if (list.head == list.tail) {
        list.head = list.tail = nullptr;
    } else {
        list.head = list.head->next;
        list.head->prev = nullptr;
    }
    
    delete temp;
    list.size--;
}

void LREMOVE_TAIL(DoublyList& list) {
    if (!list.head) {
        cout << "Список пуст!\n";
        return;
    }
    
    LNode* temp = list.tail;
    if (list.head == list.tail) {
        list.head = list.tail = nullptr;
    } else {
        list.tail = list.tail->prev;
        list.tail->next = nullptr;
    }
    
    delete temp;
    list.size--;
}

void LREMOVE_BEFORE(DoublyList& list, const string& target) {
    if (!list.head || !list.head->next) {
        cout << "Недостаточно элементов!\n";
        return;
    }
    
    LNode* curr = list.head;
    while (curr && curr->data != target) {
        curr = curr->next;
    }
    
    if (!curr) {
        cout << "Элемент " << target << " не найден!\n";
        return;
    }
    
    if (!curr->prev) {
        cout << "Нет элемента до " << target << "!\n";
        return;
    }
    
    LNode* toRemove = curr->prev;
    
    if (toRemove == list.head) {
        list.head = curr;
        list.head->prev = nullptr;
    } else {
        toRemove->prev->next = curr;
        curr->prev = toRemove->prev;
    }

    delete toRemove;
    list.size--;
}

void LREMOVE_AFTER(DoublyList& list, const string& target) {
    if (!list.head || !list.head->next) {
        cout << "Недостаточно элементов!\n";
        return;
    }
    
    LNode* curr = list.head;
    while (curr && curr->data != target) {
        curr = curr->next;
    }
    
    if (!curr) {
        cout << "Элемент " << target << " не найден!\n";
        return;
    }
    
    if (!curr->next) {
        cout << "Нет элемента после " << target << "!\n";
        return;
    }
    
    LNode* toRemove = curr->next;
    
    if (toRemove == list.tail) {
        list.tail = curr;
        list.tail->next = nullptr;
    } else {
        curr->next = toRemove->next;
        toRemove->next->prev = curr;
    }
    
    delete toRemove;
    list.size--;
}

bool LREMOVE_BY_VALUE(DoublyList& list, const string& value) {
    LNode* curr = list.head;
    while (curr) {
        if (curr->data == value) {
            if (curr == list.head) {
                list.head = curr->next;
                if (list.head) list.head->prev = nullptr;
            } else if (curr == list.tail) {
                list.tail = curr->prev;
                if (list.tail) list.tail->next = nullptr;
            } else {
                curr->prev->next = curr->next;
                curr->next->prev = curr->prev;
            }
            
            delete curr;
            list.size--;
            return true;
        }
        curr = curr->next;
    }
    
    cout << "Элемент " << value << " не найден!\n";
    return false;
}

bool LFIND(const DoublyList& list, const string& value) {
    LNode* curr = list.head;
    int position = 0;
    while (curr) {
        if (curr->data == value) {
            cout << "Элемент " << value << " найден на позиции " << position << endl;
            return true;
        }
        curr = curr->next;
        position++;
    }
    cout << "Элемент " << value << " не найден!\n";
    return false;
}

void LPRINT_FORWARD(const DoublyList& list) {
    if (!list.head) { 
        cout << "Список пуст!\n"; 
        return; 
    }
    cout << "Двусвязный вперед (" << list.size << "): ";
    LNode* curr = list.head;
    while (curr) {
        cout << curr->data;
        if (curr->next) cout << " <-> ";
        curr = curr->next;
    }
    cout << endl;
}

void LPRINT_BACKWARD(const DoublyList& list) {
    if (!list.head) { 
        cout << "Список пуст!\n"; 
        return; 
    }
    cout << "Двусвязный назад (" << list.size << "): ";
    LNode* curr = list.tail;
    while (curr) {
        cout << curr->data;
        if (curr->prev) cout << " <-> ";
        curr = curr->prev;
    }
    cout << endl;
}

void LPRINT(const DoublyList& list) {
    LPRINT_FORWARD(list);
}

void LCLEAR(DoublyList& list) {
    while (list.head) {
        LNode* temp = list.head;
        list.head = list.head->next;
        delete temp;
    }
    list.tail = nullptr;
    list.size = 0;
    cout << "Список очищен" << endl;
}

File: LR1_GO/DoublyList.go
package main

import "fmt"

func LADD_HEAD(list *DoublyList, value string) {
	newNode := &LNode{Data: value, Next: list.Head, Prev: nil}
	
	if list.Head == nil {
		list.Head = newNode
		list.Tail = newNode
	} else {
		newNode.Next = list.Head
		list.Head.Prev = newNode
		list.Head = newNode
	}
	list.Size++
}

func LADD_TAIL(list *DoublyList, value string) {
	newNode := &LNode{Data: value, Next: nil, Prev: list.Tail}
	
	if list.Head == nil {
		list.Head = newNode
		list.Tail = newNode
	} else {
		newNode.Prev = list.Tail
		list.Tail.Next = newNode
		list.Tail = newNode
	}
	list.Size++
}

func LADD_BEFORE(list *DoublyList, value string, target string) {
	if list.Head == nil {
		fmt.Println("Список пуст!")
		return
	}
	
	curr := list.Head
	for curr != nil && curr.Data != target {
		curr = curr.Next
	}
	
	if curr == nil {
		fmt.Printf("Элемент %s не найден!\n", target)
		return
	}
	
	newNode := &LNode{Data: value, Next: curr, Prev: curr.Prev}
	
	if curr == list.Head {
		newNode.Next = list.Head
		list.Head.Prev = newNode
		list.Head = newNode
	} else {
		newNode.Prev = curr.Prev
		newNode.Next = curr
		curr.Prev.Next = newNode
		curr.Prev = newNode
	}
	list.Size++
}

func LADD_AFTER(list *DoublyList, value string, target string) {
	if list.Head == nil {
		fmt.Println("Список пуст!")
		return
	}
	
	curr := list.Head
	for curr != nil && curr.Data != target {
		curr = curr.Next
	}
	
	if curr == nil {
		fmt.Printf("Элемент %s не найден!\n", target)
		return
	}
	
	newNode := &LNode{Data: value, Next: curr.Next, Prev: curr}
	
	if curr == list.Tail {
		newNode.Prev = list.Tail
		list.Tail.Next = newNode
		list.Tail = newNode
	} else {
		newNode.Prev = curr
		newNode.Next = curr.Next
		curr.Next.Prev = newNode
		curr.Next = newNode
	}
	list.Size++
}

func LREMOVE_HEAD(list *DoublyList) {
	if list.Head == nil {
		fmt.Println("Список пуст!")
		return
	}
	
	if list.Head == list.Tail {
		list.Head = nil
		list.Tail = nil
	} else {
		list.Head = list.Head.Next
		list.Head.Prev = nil
	}
	list.Size--
}

func LREMOVE_TAIL(list *DoublyList) {
	if list.Head == nil {
		fmt.Println("Список пуст!")
		return
	}
	
	if list.Head == list.Tail {
		list.Head = nil
		list.Tail = nil
	} else {
		list.Tail = list.Tail.Prev
		list.Tail.Next = nil
	}
	list.Size--
}

func LREMOVE_BEFORE(list *DoublyList, target string) {
	if list.Head == nil || list.Head.Next == nil {
		fmt.Println("Недостаточно элементов!")
		return
	}
	
	curr := list.Head
	for curr != nil && curr.Data != target {
		curr = curr.Next
	}
	
	if curr == nil {
		fmt.Printf("Элемент %s не найден!\n", target)
		return
	}
	
	if curr.Prev == nil {
		fmt.Printf("Нет элемента до %s!\n", target)
		return
	}
	
	toRemove := curr.Prev
	
	if toRemove == list.Head {
		list.Head = curr
		list.Head.Prev = nil
	} else {
		toRemove.Prev.Next = curr
		curr.Prev = toRemove.Prev
	}
	list.Size--
}

func LREMOVE_AFTER(list *DoublyList, target string) {
	if list.Head == nil || list.Head.Next == nil {
		fmt.Println("Недостаточно элементов!")
		return
	}
	
	curr := list.Head
	for curr != nil && curr.Data != target {
		curr = curr.Next
	}
	
	if curr == nil {
		fmt.Printf("Элемент %s не найден!\n", target)
		return
	}
	
	if curr.Next == nil {
		fmt.Printf("Нет элемента после %s!\n", target)
		return
	}
	
	toRemove := curr.Next
	
	if toRemove == list.Tail {
		list.Tail = curr
		list.Tail.Next = nil
	} else {
		curr.Next = toRemove.Next
		toRemove.Next.Prev = curr
	}
	list.Size--
}

func LREMOVE_BY_VALUE(list *DoublyList, value string) bool {
	curr := list.Head
	for curr != nil {
		if curr.Data == value {
			if curr == list.Head {
				list.Head = curr.Next
				if list.Head != nil {
					list.Head.Prev = nil
				}
			} else if curr == list.Tail {
				list.Tail = curr.Prev
				if list.Tail != nil {
					list.Tail.Next = nil
				}
			} else {
				curr.Prev.Next = curr.Next
				curr.Next.Prev = curr.Prev
			}
			list.Size--
			return true
		}
		curr = curr.Next
	}
	
	fmt.Printf("Элемент %s не найден!\n", value)
	return false
}

func LFIND(list *DoublyList, value string) bool {
	curr := list.Head
	position := 0
	for curr != nil {
		if curr.Data == value {
			fmt.Printf("Элемент %s найден на позиции %d\n", value, position)
			return true
		}
		curr = curr.Next
		position++
	}
	fmt.Printf("Элемент %s не найден!\n", value)
	return false
}

func LPRINT_FORWARD(list *DoublyList) {
	if list.Head == nil {
		fmt.Println("Список пуст!")
		return
	}
	
	fmt.Printf("Двусвязный вперед (%d): ", list.Size)
	curr := list.Head
	for curr != nil {
		fmt.Print(curr.Data)
		if curr.Next != nil {
			fmt.Print(" <-> ")
		}
		curr = curr.Next
	}
	fmt.Println()
}

func LPRINT_BACKWARD(list *DoublyList) {
	if list.Head == nil {
		fmt.Println("Список пуст!")
		return
	}
	
	fmt.Printf("Двусвязный назад (%d): ", list.Size)
	curr := list.Tail
	for curr != nil {
		fmt.Print(curr.Data)
		if curr.Prev != nil {
			fmt.Print(" <-> ")
		}
		curr = curr.Prev
	}
	fmt.Println()
}

func LPRINT(list *DoublyList) {
	LPRINT_FORWARD(list)
}

func LCLEAR(list *DoublyList) {
	list.Head = nil
	list.Tail = nil
	list.Size = 0
	fmt.Println("Список очищен")
}
